mirascope/beta/realtime/openai/realtime.py

```diff
@@ -0,0 +1,255 @@
+                    case "response.audio_transcript.done":
+                        await self._process_specific_receiver(
+                            audio_transcript_chunk, "audio_transcript"
+                        )
+                        audio_transcript_chunk = ""
+                    case "response.audio_transcript.delta":
+                        audio_transcript_chunk += message["delta"]
+                        await self._process_specific_receiver(
+                            audio_transcript_chunk, "audio_transcript_chunk"
+                        )
+
+                    # TODO: Add more cases like "response.function_call_arguments.done"
+
+                # Call the receivers for all messages
+                for receiver in self.receivers[None]:
+                    await receiver(message, self.context)
+        except asyncio.CancelledError:
+            print("Receiver has been cancelled")
+        finally:
+            print("Receiver has been closed")
+
+    async def run(self) -> None:
+        if self.is_running():
+            raise RuntimeError("Already running")
+
+        headers = {
+            "Authorization": f"Bearer {os.environ["OPENAI_API_KEY"]}",
+            "OpenAI-Beta": "realtime=v1",
+        }
+
+        async for conn in connect(
+            uri=f"wss://api.openai.com/v1/realtime?model={self.model}",
+            additional_headers=headers,
+        ):
+            try:
+                receiver_task = asyncio.create_task(self._process_receiver(conn))
+                sender_task = asyncio.create_task(self._process_sender(conn))
+                await asyncio.gather(receiver_task, sender_task)
+            except websockets.ConnectionClosed:
+                continue
+
+    def is_running(self) -> bool:
+        return self._connection is not None
+
+
+if __name__ == "__main__":
+    app = OpenAIRealtime(
+        "gpt-4o-realtime-preview-2024-10-01",
+        {"context": "context"},
+        **{"vda_mode": True},
+    )
+
+    @app.receiver("text")
+    async def receive_text(response: str, context: dict[str, Any]) -> None:
+        print(f"AI(text): {response}")
+
+    @app.receiver("audio")
+    async def receive_audio(response: AudioSegment, context: dict[str, Any]) -> None:
+        play(response)
+
+    @app.receiver("audio_transcript")
+    async def receive_audio_transcript(response: str, context: dict[str, Any]) -> None:
+        print(f"AI(audio_transcript): {response}")
+
+    #
+    # @app.sender()
+    # async def send_message(context: dict[str, Any]) -> str:
+    #     while True:
+    #         print("Enter your message: ", end="", flush=True)
+    #         message = await asyncio.to_thread(input, )
+    #         return message
+    #
+    @app.sender()
+    async def send_audio_as_stream(
+        context: dict[str, Any],
+    ) -> AsyncGenerator[BytesIO, None]:
+        message = await async_input(
```

Comment:
how do we enable the user to constantly send/receive audio for a more realtime conversation? i.e. vda mode where the assistant can interrupt the user by detecting when they are done automatically? i.e. the user doesn't do anything except stop talking?